src/wallet/coinselection.cpp

```diff
@@ -452,12 +452,16 @@ void SelectionResult::AddInputs(const std::set<COutput>& inputs, bool subtract_f
 
 void SelectionResult::Merge(const SelectionResult& other)
 {
+    // Obtain the expected selected inputs count after the merge (for now, duplicates are not allowed)
+    const size_t expected_count = m_selected_inputs.size() + other.m_selected_inputs.size();
+
     m_target += other.m_target;
     m_use_effective |= other.m_use_effective;
     if (m_algo == SelectionAlgorithm::MANUAL) {
         m_algo = other.m_algo;
     }
     util::insert(m_selected_inputs, other.m_selected_inputs);
+    assert(m_selected_inputs.size() == expected_count);
```

Comment:
It's a bit unclear to me how the assert added in eb76557331b4f34b568da1d9c589c7005d7ab46d is more safe? Wouldn't this mean we still crash if we haven't caught a similar bug before shipping?

Reply:
I suggest removing this last commit for this PR, and following up with an "add sanity checks to coin selection code" PR.

Of course we don't want to create a bad transaction, but an assert would cause the user's node to crash in case we have implemented it wrong. I think something better could be to add `if(!Assume(...)) {return "sorry, something went wrong and transaction creation failed. please report this bug...";}` which should stop the wallet from doing something horrible but also not crash. Apologies if this is bikesheddy - my concrete suggestion is to remove this for now, and do a followup PR for these types of changes. Other appropriate sanity checks could include `AddInput()` never adds an input that's already there, `SelectCoins` result should cover `selection_target`, etc.

Reply:
> Of course we don't want to create a bad transaction, but an assert would cause the user's node to crash in case we have implemented it wrong. I think something better could be to add if(!Assume(...)) {return "sorry, something went wrong and transaction creation failed. please report this bug...";} which should stop the wallet from doing something horrible but also not crash.

Sure. I totally agree on using more `Assume` + user friendly error return messages than crashing the node with a bare assertion. 

I do actually implemented the capability to return error messages from the Coin Selection process inside [#25806](https://github.com/bitcoin/bitcoin/pull/25806) -> 45bc10687f7b0131fd1c4575406b62f0608e2c01 few months ago (which is later used to return specific error messages instead of the current general "Insufficient Funds". Which is required to implement your suggestions as we currently don't have a way to return specific errors to the upper layers).

So, in short, could drop the last commit and open a follow-up PR moving forward with it, and more, for sure, but.. as I added the last commit by reviewers request, would be good to see if @S3RK is sync with this reasoning too (otherwise we would be going in circles).
Note: Since, after the fix, this problem cannot realistically happen anymore, I'm fine to move either way.

Reply:
I think it's find to change this to `Assume` for this PR. We should probably try to avoid asserts in general.

Reply:
I'll be fine with any of the following: keep asserts as I proposed, replaced them with `Assume` for this PR, remove the commit and address this in a follow up. With that said, here's my strong opinion held lightly.

In the absence of ability of returning error messages from coin selection we have only two options. So the question is what is better within the scope of this PR:
a) crash the node if assert is hit
b) keep executing and potentially creating bad transactions.

Both are bad, but a) is strictly better, because recovering from a crash is much easier than from a bad tx. Replacing `assert`s and returning error messages could/should done in a follow up.

Now I agree that in general the wallet should not ever crash the node, but for this we need a general mechanism, something like multiprocess.

Also, my original suggestion included check within `AddInput()` and `AddInputs()`, as `SelectionResult` is not designed to work with duplicated inputs. This would make the class safe and its interface consistent.

Reply:
Another alternative (not sure if applicable here) would be to throw an exception. This should only crash when running in the gui, but not when called from the RPC server.

Reply:
It seems we all (mostly) agree that the `assert` is not ideal and how to deal with this should be addressed in a follow-up with more discussion, so I'd recommend dropping the commit so we don't merge code that we know will be removed/refactored shortly after and to avoid any confusion when backporting commits from this PR for v24

Reply:
working on it 🚜 .